### PR TITLE
withDecay - add rubber band effect

### PR DESCRIPTION
**Pull request by piaskowyk**
## Description

I added a new parameter to `withDecay` animation:

#### Before
https://user-images.githubusercontent.com/36106620/168560831-0ccb4092-50a1-491c-8f75-65b0cb0155d8.mov

#### After
https://user-images.githubusercontent.com/36106620/168560869-a277860d-b2ec-42c6-8841-3ed89785fede.mov

## How to use it
```js
withDecay({
  clamp: [10, 100],
  velocity: 200,
  rubberBandEffect: true, // <-- flag
  rubberBandFactor: 0.6, // <-- optional config
});
```

## Example

<details>
<summary>Example code</summary>

```js
import Animated, { useSharedValue, withDecay, useAnimatedStyle } from 'react-native-reanimated';
import { StyleSheet, Button, View } from 'react-native';

import React from 'react';

export default function ScrollViewExample() {
  const position = useSharedValue(0.000001);
  const style = useAnimatedStyle(() => {
    return {
      transform: [{
        translateX: position.value
      }]
    }
  });
  return (
    <Animated.View style={{marginTop: 50, display: 'flex', alignItems: 'center'}}>
      <Button title='right' onPress={() => {
        position.value = withDecay({
          clamp: [0.000001, 100],
          velocity: 200,
          rubberBandEffect: true,
        });
      }} />
      <Button title='left' onPress={() => {
        position.value = withDecay({
          clamp: [0.000001, 100],
          velocity: -200,
          rubberBandEffect: true,
        });
      }} />
      <View style={{width: 200, height: 150, backgroundColor: 'red'}}>
        <Animated.View style={[{width: 100, height: 100, backgroundColor: 'blue'}, style]} />
      </View>
    </Animated.View>
  );
}

```

</details>

### CODE REVIEW

File: src/reanimated2/animation/decay.ts [hunk 1]
```diff
@@ -59,40 +62,81 @@ export function withDecay(
...
+          lastTimestamp,
+          startTimestamp,
+          current,
+          initialVelocity,
+          velocity,
+        } = animation;
+
+        const deltaTime = Math.min(now - lastTimestamp, 64);
+        const clampIndex = initialVelocity > 0 ? 1 : 0;
+        let deriveration = 0;
```
Comment (graszka22) [inline src/reanimated2/animation/decay.ts:79]: What is "deriveration"? Maybe use simply something like "distance" or you wanted to use "deviation"?

**Review by graszka22 — APPROVED** [review]
LGTM, one comment about spelling

**Comment by wojtus7** [comment]
Unfortunately, the animation curve is kinda harsh, and the softClampFactor doesn't give very much control of how the end animation looks like :(

File: src/reanimated2/animation/decay.ts (hunk 1 above)
Comment (tomekzaw) [reply src/reanimated2/animation/decay.ts:79]: Or "derivative"?

File: src/reanimated2/animation/decay.ts [hunk 2]
```diff
@@ -157,8 +157,10 @@ export function withDecay(
           `config.velocityFactor must be greather then 0 but is ${config.velocityFactor}`
         );
       }
-      if (config.softClamp && !config.clamp) {
-        throw Error(`softClamp requires clamp array`);
+      if (config.rubberBandEffect && !config.clamp) {
+        throw Error(
+          `you need to set clamp property when you want to use the rubberBandEffect`
```
Comment (tomekzaw) [inline src/reanimated2/animation/decay.ts:162]:
```suggestion
          'You need to set `clamp` property when using `rubberBandEffect`.'
```

**Review by tomekzaw — APPROVED** [review]
🥇

**Comment by alantoa** [comment]
Hi guys, I very love this feature, can't wait to use it!
I wonder when it will be released?

**Comment by smontlouis** [comment]
agreed with @wojtus7 

Just tried it and the effect is way to stiff

**Comment by zigcccc** [comment]
was this removed? The docs still refer to it, but it seems like it's no longer there?

**Pull request merged**